### ISSUE TITLE
Document interaction of human-readable rule names and error handling

### ISSUE THREAD

**Issue by sampsyo** [pegjs/pegjs#368]
The presence of a human-readable name for a rule (regardless of its content) can affect the kinds of errors that PEG.js produces. This is a subtle effect that is important for getting good parse errors, but it doesn't seem to be documented.

Here's a tiny grammar that parses sequences of lower-case letters separated by semicolons:

```
S = A (";" A)*
A = [a-z]
```

What happens if I mistakenly use an upper-case letter? If I try to parse `a;B`, this grammar emits the error:

```
Expected [a-z] but "B" found.
```

But if I assign a human-readable name for the first rule:

```
S "foo" = A (";" A)*
A = [a-z]
```

I get a very different error message:

```
Expected end of input but ";" found.
```

The parser seems to prefer "blaming" named rules over unnamed rules. This seems like a reasonable way to control error messages, and there's kind of an intuitive logic to it: unnamed rules are more likely to be some kind of internal implementation detail, and the library tries to avoid exposing them too much to the user.

But unless I'm missing something, this behavior isn't documented. (I found it by trial and error.) I'd be happy to draft some text to help document this if I can have some help with understanding exactly how this feature works.

**Comment by dmajda** [comment]
@sampsyo You are right that this behavior is not documented, or rather documented quite vaguely:

> A rule can also contain *human-readable name* that is used in error messages (in our example, only the `integer` rule has a human-readable name).

In fact, the whole error handling area is currently not documented much.

> I'd be happy to draft some text to help document this if I can have some help with understanding exactly how this feature works.

Are you still interested in this? You pretty much already discovered how the feature works — low-level expressions (string literals, character classes, and `.`) produce *expectations* on match failures and a rule with human-readable name just overwrites these expectations with a one containing rule’s human-readable name.

**Comment by sampsyo** [comment]
Sure! I won't pretend that I understand it completely, but I can try putting together a few examples to demonstrate the effect. I'll open a PR.

**Comment by futagoza** [comment]
Merged #415
